pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,630 @@
+		}
+		var allPods []*v1.Pod
+		for _, ns := range namespaces {
+			pods, err := c.podLister.Pods(ns.Name).List(podSelector)
+			if err != nil {
+				return nil, err
+			}
+			allPods = append(allPods, pods...)
+		}
+		return allPods, nil
+	}
+	return c.podLister.List(podSelector)
+}
+
+func genPortNameUDPTunnel(tunnel *v1alpha2.UDPTunnel) string {
+	hash := sha1.New() // #nosec G401: not used for security purposes
+	hash.Write(net.ParseIP(tunnel.RemoteIP))
+	if tunnel.DestinationPort != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.DestinationPort))))
+	}
+	if tunnel.VNI != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.VNI))))
+	}
+	return hex.EncodeToString(hash.Sum(nil))[:6]
+}
+
+func genPortNameGRETunnel(tunnel *v1alpha2.GRETunnel) string {
+	hash := sha1.New() // #nosec G401: not used for security purposes
+	hash.Write(net.ParseIP(tunnel.RemoteIP))
+	if tunnel.Key != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.Key))))
+	}
+	return hex.EncodeToString(hash.Sum(nil))[:6]
+}
+
+func genPortNameERSPANTunnel(tunnel *v1alpha2.ERSPANTunnel) string {
+	hash := sha1.New() // #nosec G401: not used for security purposes
+	hash.Write(net.ParseIP(tunnel.RemoteIP))
+	if tunnel.SessionID != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.SessionID))))
+	}
+	if tunnel.Index != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.Index))))
+	}
+	if tunnel.Dir != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.Dir))))
+	}
+	if tunnel.HardwareID != nil {
+		hash.Write([]byte(strconv.Itoa(int(*tunnel.HardwareID))))
+	}
+	return hex.EncodeToString(hash.Sum(nil))[:6]
+}
+
+func (c *Controller) getDevicePort(device *v1alpha2.TrafficControlPort) (ofPort uint32, err error) {
+	if device == nil {
+		return 0, fmt.Errorf("device is invalid")
+	}
+	var ofPortTmp int32
+	createUDPTunnel := func(tunnelType ovsconfig.TunnelType, portName, remoteIP string, dstPort int32) error {
+		externalIDs := map[string]interface{}{
+			interfacestore.AntreaInterfaceTypeKey: interfacestore.AntreaTunnel,
+		}
+		if dstPort != 0 {
+			externalIDs["dst_port"] = strconv.Itoa(int(dstPort))
+		}
+		c.ovsPortUpdateMutex.Lock()
+		defer c.ovsPortUpdateMutex.Unlock()
+		portUUID, err := c.ovsBridgeClient.CreateTunnelPortExt(
+			portName, tunnelType, 0, false, "", remoteIP, "", externalIDs)
+		if err != nil {
+			return err
+		}
+		ofPortTmp, err = c.ovsBridgeClient.GetOFPort(portName, false)
+		if err != nil {
+			return err
+		}
+		itf := interfacestore.NewTunnelInterface(portName, tunnelType, nil, false)
+		itf.OVSPortConfig = &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: ofPortTmp}
+		c.interfaceStore.AddInterface(itf)
+		ofPort = uint32(ofPortTmp)
+		return nil
+	}
+	switch {
+	case device.OVSInternal != nil:
+		if ofPort, err := c.ovsBridgeClient.GetOFPort(device.OVSInternal.Name, false); err == nil {
+			return uint32(ofPort), nil
+		}
+		externalIDs := map[string]interface{}{
+			interfacestore.AntreaInterfaceTypeKey: interfacestore.AntreaUnset,
```

Comment:
Can we add a type specific to traffic-control so that we know the interface is auto-created and we can remove it when it's no longer used?

Reply:
👍 good idea